### ISSUE TITLE
Oracle sometimes gives the same bind a different bind #number.

### ISSUE THREAD

**Issue by NormanDunbar** [NormanDunbar/TraceMiner2#6]
Consider the following code, it has 3 different binds in it:

* ``:inOutCnt``
* ``:trimLines``
* ``:outLines``

However, ``:inOutCnt`` is used twice. The trace file should have a "no ocadef found" for a ``BIND #3`` entry, but it doesn't!

TraceMiner2 is searching for a bind that already exists with the same bind id as well as name, it needs to check for a bind with the name only, just in case Oracle is playing silly beggars again!

Extract from a trace file showing the problem follows:

```
PARSING IN CURSOR #4573830864 len=1512 dep=0 uid=220 oct=47 lid=220 tim=1813092665441 hv=449603905 ad='7000100243e40e0' sqlid='1v0a01hdcsua1'
DECLARE
   v_lines    SYS.DBMS_OUTPUT.chararr;
   v_line     VARCHAR2 (32767);
   v_clob     CLOB;
   v_cnt      NUMBER;
   v_output   VARCHAR2 (32767);
   PROCEDURE write_to_clob (p_clob IN OUT CLOB, p_str IN OUT VARCHAR2)
   AS
   BEGIN
      BEGIN
         DBMS_LOB.writeappend (p_clob, LENGTH (p_str), p_str);
      EXCEPTION
         WHEN VALUE_ERROR
         THEN
            NULL;
      END;
      p_str := NULL;
   END;
BEGIN
   v_cnt := :inOutCnt;
   SYS.DBMS_OUTPUT.get_lines (v_lines, v_cnt);
   DBMS_LOB.createtemporary (v_clob, FALSE, DBMS_LOB.session);
   DBMS_LOB.open (v_clob, DBMS_LOB.lob_readwrite);
   v_output := NULL;
   FOR line IN 1 .. v_lines.COUNT
   LOOP
      IF :trim_line = 1 THEN
         v_line := TRIM (v_lines (line));
      ELSE
         v_line := v_lines (line);
      END IF;
      IF v_line IS NOT NULL
      THEN
         v_line := REPLACE (v_line, CHR (0));
         IF LENGTHB (v_output) + LENGTHB (v_line) > 32767
         THEN
            write_to_clob (v_clob, v_output);
         END IF;
         v_output := v_output || v_line;
         IF line < v_lines.COUNT
         THEN
            IF LENGTHB (v_output) + 2 > 32767
            THEN
               write_to_clob (v_clob, v_output);
            END IF;
            v_output := v_output || CHR (13) || CHR (10);
         END IF;
      END IF;
   END LOOP;
   write_to_clob (v_clob, v_output);
   if v_cnt > 0 then
     :outLines := v_clob;
   end if;
   :inOutCnt := v_cnt;
   DBMS_LOB.freetemporary (v_clob);
END;
END OF STMT
PARSE #4573830864:c=49,e=49,p=0,cr=0,cu=0,mis=0,r=0,dep=0,og=1,plh=0,tim=1813092665440
```

The bind data for the execution of the above is as follows:

```
BINDS #4573830864:
 Bind#0
  oacdty=02 mxl=22(22) mxlc=00 mal=00 scl=00 pre=00
  oacflg=01 fl2=1000000 frm=00 csi=00 siz=48 off=0
  kxsbbbfp=110a59528  bln=22  avl=02  flg=05
  value=500
 Bind#1
  oacdty=02 mxl=22(22) mxlc=00 mal=00 scl=00 pre=00
  oacflg=01 fl2=1000000 frm=00 csi=00 siz=0 off=24
  kxsbbbfp=110a59540  bln=22  avl=02  flg=01
  value=1
 Bind#2
  oacdty=112 mxl=4000(4000) mxlc=00 mal=00 scl=00 pre=00
  oacflg=01 fl2=1000000 frm=01 csi=46 siz=4000 off=0
  kxsbbbfp=110aa9d88  bln=4000  avl=00  flg=05
WAIT #4573830864: nam='SQL*Net message to client' ela= 3 driver id=1413697536 #bytes=1 p3=0 obj#=-1 tim=1813092666558
EXEC #4573830864:c=510,e=509,p=0,cr=0,cu=0,mis=0,r=1,dep=0,og=1,plh=0,tim=1813092666582
WAIT #4573830864: nam='SQL*Net message from client' ela= 34872 driver id=1413697536 #bytes=1 p3=0 obj#=-1 tim=1813092701519
CLOSE #4573830864:c=19,e=20,dep=0,type=3,tim=1813092701597
=====================
```

As you can see, it doesn't mention the fact that ``Bind #3`` is even present - whihc could be because this is PL/SQL? Maybe? Perhaps?

**Comment by NormanDunbar** [comment]
This looks to have been fixed by the solution to issue #9, but I'll leave the issue open until I am certain that it has been fully resolved.

**Comment by NormanDunbar** [comment]
Nope, #9 did not fix #6 after all. At least, not in the example given above.
Still working on it though. :-)

**Comment by NormanDunbar** [comment]
It looks like it's fixed. At least, in the trace files I've been able to test. Commit d9c42c5 does the job.
If a bind appears in the SQL, but not in the binds for the cursor, use its name rather than its value in the substitution at exec time.
The comment in commit d9c42c5 should reference this issue, not issue #5.
